docs/dev/developer-guides/transactions/fee-model.md

```diff
@@ -1,72 +1,104 @@
+## Gas estimation for transactions
 
-Since the costs for publishing the calldata on L1 are very volatile, the number of `gas` needed for changing a storage slot is not constant. For each block, the operator defines the following dynamic parameters:
+Ethereum has a constant of 21000 gas that covers the intrinsic costs of processing a transaction, i.e. checking the signature and updating the nonce for the account.
 
-- `gas_price` — the table for the current base price in each token. The value of this parameter is used to determine the costs of VM execution in each token.
-- `gas_per_pubdata` — the price in `gas` for publishing one byte of data to Ethereum.
+On zkSync Era this varies because we support custom and paymaster accounts. These accounts require a (usually) higher amount of gas than EOAs. zkSync Era provides functions for estimating the cost of a transaction regardless of the type of account.
 
-**Please note that the public data is published only for state diffs.** If the same storage slot is updated 10 times in the same rollup block, only the final update will be published on Ethereum, thus only charging for public data once.
+The transaction fee estimate depends on the entire transaction flow, including validation and execution. The `eth_estimateGas` function uses binary search to find the smallest gas value under which the transaction succeeds.
 
-## Fee model: zkSync vs Ethereum
+For any Rust developers interested in the zkSync Era implementation for gas estimation, see the Rust code in our repo.
 
-We want to show the clear distinction between our fee model and the Ethereum one.
-Also, unlike Ethereum, where most of the opcodes have very distinct gas prices, in zkSync the cost of the opcodes is different, hence zkEVM opcodes will likely have similar `gas` prices.
-Generally, the execution itself (arithmetic operations, which do not involve storage updates) is very cheap. As in Ethereum, most of the cost is incurred for storage updates.
+### Transaction length
 
-## What does this mean to me?
+zkSync Era publishes state diffs on-chain. The cost of the transaction, however, may still depend on transaction length because the sequencer stores long transactions in-memory.
 
-Despite the differences, the fee model is quite similar to the one of Ethereum; the most costly operation is the storage change. One of the advantages of zk rollups over-optimistic rollups is that, instead of publishing all the transaction data, zk rollups can publish only state diffs, thus making fewer storage changes.
+Long transactions incur additional costs during interactions with an account. zkSync Era works with many types of accounts and so we cannot make assumptions about signature length. Furthermore, given that a signature (and thus its length) is unavailable at the time of fee estimation, we cannot precisely estimate the cost of such a transaction. To mitigate this, we multiply the recommended cost of the transaction by a small percentage.
```

Comment:
```suggestion
Long transactions incur additional costs during interactions with an account. zkSync Era works with differet types of accounts and it cannot make assumptions about signature length. Furthermore, given that a signature (and thus its length) is unavailable at the time of fee estimation, we cannot precisely estimate the cost of such a transaction. To mitigate this, we multiply the recommended cost of the transaction by a small percentage.
```
There are really only 2 types: EOAs and AA ("smart accounts")

Reply:
Let's avoid "we" whenever possible: I think we can use "the protocol" here.

Reply:
I like we, it's inclusive ..

Reply:
Once the sequencer becomes decentralised, there would not be a "we"...

Reply:
we means everyone, for full decentralization there is only "we", but it's not a big deal, the protocol is probably better

Reply:
go ahead and suggest again and i'll commit